compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -540,6 +543,8 @@ pub struct InferCtxtBuilder<'tcx> {
     defining_use_anchor: DefiningAnchor,
     considering_regions: bool,
     fresh_typeck_results: Option<RefCell<ty::TypeckResults<'tcx>>>,
+    normalize_fn_sig_for_diagnostic:
+        Option<Lrc<dyn Fn(&InferCtxt<'_, 'tcx>, ty::PolyFnSig<'tcx>) -> ty::PolyFnSig<'tcx>>>,
```

Comment:
i am somewhat unhappy about this.

i think we should yeet all of the type error reporting out of `rustc_infer` and put them into a separate crate. Now, that is a lot more complex than just adding a single field and would require an MCP, but it should have a pretty positive impact:
- faster recompilation times when working on type diagnostics
- a clear separation between ("this may be useful everywhere" and "this is only for diags")
- errors can use stuff like normalization without this much hassle.

Ideally we do the same for any errors in `rustc_trait_selection` but I know that doing that will be pretty annoying/difficult, so maybe not :grin:

I am not strictly opposed to that change, as the alternative is a lot more effort, and we shouldn't block useful things because of something which may never happen. I would be up to mentor/review this approach though if you're interested though :grin: